[Bug] an exception occurred when restore job from a snapshot which previously sync multiple tables and set scan.incremental.close-idle-reader.enabled=true

### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.


### Flink version

1.7.1

### Flink CDC version

2.4.0

### Database and its version

MySQL 5.7

### Minimal reproduce step

1. start a cdc job wich syncs multiple tables, and has a parallelism more than 1, some configs include:
```
scan.newly-added-table.enabled=true
scan.incremental.close-idle-reader.enabled=true
```
3. stop with a savepoint
4. restart the job with savepoint, add a new table and do not change the previous parallelism
5. an exception occurred:Can not restore vertex Source: MySQL Source -> Process -> (test: Writer -> test: Committer, test2: Writer -> test2: Committer)(cbc357ccb763df2852fee8c4fc7d55f2) which contain mixed operator finished state: [ALL_RUNNING, PARTIALLY_FINISHED]


### What did you expect to see?

The job syncs all the tables, including newly added tables.

### What did you see instead?

An Exception occurred:
org.apache.flink.runtime.client.JobInitializationException: Could not start the JobMaster.
	at org.apache.flink.runtime.jobmaster.DefaultJobMasterServiceProcess.lambda$new$0(DefaultJobMasterServiceProcess.java:97)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1595)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.CompletionException: org.apache.flink.util.FlinkRuntimeException: Can not restore vertex Source: MySQL Source -> Process -> (test: Writer -> test: Committer, test2: Writer -> test2: Committer)(cbc357ccb763df2852fee8c4fc7d55f2) which contain mixed operator finished state: [ALL_RUNNING, PARTIALLY_FINISHED]
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1592)
	... 3 more
Caused by: org.apache.flink.util.FlinkRuntimeException: Can not restore vertex Source: MySQL Source -> Process -> (test: Writer -> test: Committer, test2: Writer -> test2: Committer)(cbc357ccb763df2852fee8c4fc7d55f2) which contain mixed operator finished state: [ALL_RUNNING, PARTIALLY_FINISHED]
	at org.apache.flink.runtime.checkpoint.VertexFinishedStateChecker$VerticesFinishedStatusCache.calculateFinishedState(VertexFinishedStateChecker.java:190)
	at org.apache.flink.runtime.checkpoint.VertexFinishedStateChecker$VerticesFinishedStatusCache.lambda$getOrUpdate$0(VertexFinishedStateChecker.java:171)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1126)
	at org.apache.flink.runtime.checkpoint.VertexFinishedStateChecker$VerticesFinishedStatusCache.getOrUpdate(VertexFinishedStateChecker.java:169)
	at org.apache.flink.runtime.checkpoint.VertexFinishedStateChecker.validateOperatorsFinishedState(VertexFinishedStateChecker.java:70)
	at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.restoreLatestCheckpointedStateInternal(CheckpointCoordinator.java:1737)
	at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.restoreSavepoint(CheckpointCoordinator.java:1873)
	at org.apache.flink.runtime.scheduler.DefaultExecutionGraphFactory.tryRestoreExecutionGraphFromSavepoint(DefaultExecutionGraphFactory.java:223)
	at org.apache.flink.runtime.scheduler.DefaultExecutionGraphFactory.createAndRestoreExecutionGraph(DefaultExecutionGraphFactory.java:198)
	at org.apache.flink.runtime.scheduler.SchedulerBase.createAndRestoreExecutionGraph(SchedulerBase.java:365)
	at org.apache.flink.runtime.scheduler.SchedulerBase.<init>(SchedulerBase.java:210)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.<init>(DefaultScheduler.java:136)
	at org.apache.flink.runtime.scheduler.DefaultSchedulerFactory.createInstance(DefaultSchedulerFactory.java:152)
	at org.apache.flink.runtime.jobmaster.DefaultSlotPoolServiceSchedulerFactory.createScheduler(DefaultSlotPoolServiceSchedulerFactory.java:119)
	at org.apache.flink.runtime.jobmaster.JobMaster.createScheduler(JobMaster.java:371)
	at org.apache.flink.runtime.jobmaster.JobMaster.<init>(JobMaster.java:348)
	at org.apache.flink.runtime.jobmaster.factories.DefaultJobMasterServiceFactory.internalCreateJobMasterService(DefaultJobMasterServiceFactory.java:123)
	at org.apache.flink.runtime.jobmaster.factories.DefaultJobMasterServiceFactory.lambda$createJobMasterService$0(DefaultJobMasterServiceFactory.java:95)
	at org.apache.flink.util.function.FunctionUtils.lambda$uncheckedSupplier$4(FunctionUtils.java:112)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	... 3 more


### Anything else?

中文描述：
从现象上看，使用scan.incremental.close-idle-reader.enabled=true参数，作业会自动释放资源，但是释放资源后，就不能再返回最初的多并行度状态了。而实际的生产需求往往是：高并行度同步一批表 -> 释放多余资源，恢复并行度1  -> 新增同步一批表，高并行度再次同步完 -> 释放多余资源，恢复并行度1   -> 循环此过程...

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlHybridSplitAssigner.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlHybridSplitAssigner.java
@@ -189,6 +189,10 @@ public class MySqlHybridSplitAssigner implements MySqlSplitAssigner {
         snapshotSplitAssigner.close();
     }
 
+    public boolean noMoreSnapshotSplits() {
+        return snapshotSplitAssigner.noMoreSnapshotSplits();
+    }
+
     // --------------------------------------------------------------------------------------------
 
     private MySqlBinlogSplit createBinlogSplit() {

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
@@ -112,7 +112,10 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
     @Override
     public void addSplitsBack(List<MySqlSplit> splits, int subtaskId) {
         LOG.debug("The enumerator adds splits back: {}", splits);
-        if (splits.stream().anyMatch(MySqlSplit::isBinlogSplit)) {
+        Optional<MySqlSplit> binlogSplit =
+                splits.stream().filter(MySqlSplit::isBinlogSplit).findAny();
+        if (binlogSplit.isPresent()) {
+            LOG.info("The enumerator adds add binlog split back: {}", binlogSplit);
             this.binlogSplitTaskId = null;
         }
         splitAssigner.addSplits(splits);

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
@@ -75,7 +75,7 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
     private final TreeSet<Integer> readersAwaitingSplit;
     private List<List<FinishedSnapshotSplitInfo>> binlogSplitMeta;
 
-    private Integer binlogSplitTaskId;
+    @Nullable private Integer binlogSplitTaskId;
 
     public MySqlSourceEnumerator(
             SplitEnumeratorContext<MySqlSplit> context,
@@ -112,6 +112,9 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
     @Override
     public void addSplitsBack(List<MySqlSplit> splits, int subtaskId) {
         LOG.debug("The enumerator adds splits back: {}", splits);
+        if (splits.stream().anyMatch(MySqlSplit::isBinlogSplit)) {
+            this.binlogSplitTaskId = null;
+        }
         splitAssigner.addSplits(splits);
     }
 

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java
@@ -24,10 +24,12 @@ import org.apache.flink.util.FlinkRuntimeException;
 
 import org.apache.flink.shaded.guava30.com.google.common.collect.Lists;
 
+import com.ververica.cdc.connectors.mysql.source.assigners.MySqlBinlogSplitAssigner;
 import com.ververica.cdc.connectors.mysql.source.assigners.MySqlHybridSplitAssigner;
 import com.ververica.cdc.connectors.mysql.source.assigners.MySqlSplitAssigner;
 import com.ververica.cdc.connectors.mysql.source.assigners.state.PendingSplitsState;
 import com.ververica.cdc.connectors.mysql.source.config.MySqlSourceConfig;
+import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitAssignedEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitMetaEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitMetaRequestEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitUpdateAckEvent;
@@ -39,6 +41,7 @@ import com.ververica.cdc.connectors.mysql.source.events.LatestFinishedSplitsNumb
 import com.ververica.cdc.connectors.mysql.source.events.LatestFinishedSplitsNumberRequestEvent;
 import com.ververica.cdc.connectors.mysql.source.offset.BinlogOffset;
 import com.ververica.cdc.connectors.mysql.source.split.FinishedSnapshotSplitInfo;
+import com.ververica.cdc.connectors.mysql.source.split.MySqlBinlogSplit;
 import com.ververica.cdc.connectors.mysql.source.split.MySqlSplit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -71,6 +74,8 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
     // using TreeSet to prefer assigning binlog split to task-0 for easier debug
     private final TreeSet<Integer> readersAwaitingSplit;
     private List<List<FinishedSnapshotSplitInfo>> binlogSplitMeta;
+
+    private Integer binlogSplitTaskId;
 
     public MySqlSourceEnumerator(
             SplitEnumeratorContext<MySqlSplit> context,
@@ -153,6 +158,11 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
                     "The enumerator receives request from subtask {} for the latest finished splits number after added newly tables. ",
                     subtaskId);
             handleLatestFinishedSplitNumberRequest(subtaskId);
+        } else if (sourceEvent instanceof BinlogSplitAssignedEvent) {
+            LOG.info(
+                    "The enumerator receives notice from subtask {} for the binlog split assignment. ",
+                    subtaskId);
+            binlogSplitTaskId = subtaskId;
         }
     }
 
@@ -188,7 +198,10 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
                 continue;
             }
 
-            if (splitAssigner.isStreamSplitAssigned() && sourceConfig.isCloseIdleReaders()) {
+            if (splitAssigner.isStreamSplitAssigned()
+                    && sourceConfig.isCloseIdleReaders()
+                    && noMoreSnapshotSplits()
+                    && (binlogSplitTaskId != null && !binlogSplitTaskId.equals(nextAwaiting))) {
                 // close idle readers when snapshot phase finished.
                 context.signalNoMoreSplits(nextAwaiting);
                 awaitingReader.remove();
@@ -200,6 +213,9 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
             if (split.isPresent()) {
                 final MySqlSplit mySqlSplit = split.get();
                 context.assignSplit(mySqlSplit, nextAwaiting);
+                if (mySqlSplit instanceof MySqlBinlogSplit) {
+                    this.binlogSplitTaskId = nextAwaiting;
+                }
                 awaitingReader.remove();
                 LOG.info("The enumerator assigns split {} to subtask {}", mySqlSplit, nextAwaiting);
             } else {
@@ -208,6 +224,16 @@ public class MySqlSourceEnumerator implements SplitEnumerator<MySqlSplit, Pendin
                 break;
             }
         }
+    }
+
+    private boolean noMoreSnapshotSplits() {
+        if (splitAssigner instanceof MySqlHybridSplitAssigner) {
+            return ((MySqlHybridSplitAssigner) splitAssigner).noMoreSnapshotSplits();
+        } else if (splitAssigner instanceof MySqlBinlogSplitAssigner) {
+            return true;
+        }
+        throw new IllegalStateException(
+                "Unexpected subtype of MySqlSplitAssigner class when invoking noMoreSnapshotSplits.");
     }
 
     private int[] getRegisteredReader() {

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/events/BinlogSplitAssignedEvent.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/events/BinlogSplitAssignedEvent.java
@@ -1,0 +1,34 @@
+/*
+ * Copyright 2023 Ververica Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.connectors.mysql.source.events;
+
+import org.apache.flink.api.connector.source.SourceEvent;
+
+import com.ververica.cdc.connectors.mysql.source.enumerator.MySqlSourceEnumerator;
+import com.ververica.cdc.connectors.mysql.source.reader.MySqlSourceReader;
+import com.ververica.cdc.connectors.mysql.source.split.MySqlBinlogSplit;
+
+/**
+ * The {@link SourceEvent} that {@link MySqlSourceReader} sends to {@link MySqlSourceEnumerator} to
+ * notify the {@link MySqlBinlogSplit} assigned to itself.
+ */
+public class BinlogSplitAssignedEvent implements SourceEvent {
+
+    private static final long serialVersionUID = 1L;
+
+    public BinlogSplitAssignedEvent() {}
+}

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java
@@ -271,6 +271,8 @@ public class MySqlSourceReader<T>
                                     binlogSplit, sourceConfig, checkNewlyAddedTableSchema);
                     unfinishedSplits.add(mySqlBinlogSplit);
                 }
+                LOG.info(
+                        "Source reader {} received the binlog split : {}.", subtaskId, binlogSplit);
                 context.sendSourceEventToCoordinator(new BinlogSplitAssignedEvent());
             }
         }

--- a/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/reader/MySqlSourceReader.java
@@ -28,6 +28,7 @@ import org.apache.flink.util.Preconditions;
 
 import com.ververica.cdc.connectors.mysql.debezium.DebeziumUtils;
 import com.ververica.cdc.connectors.mysql.source.config.MySqlSourceConfig;
+import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitAssignedEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitMetaEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitMetaRequestEvent;
 import com.ververica.cdc.connectors.mysql.source.events.BinlogSplitUpdateAckEvent;
@@ -270,6 +271,7 @@ public class MySqlSourceReader<T>
                                     binlogSplit, sourceConfig, checkNewlyAddedTableSchema);
                     unfinishedSplits.add(mySqlBinlogSplit);
                 }
+                context.sendSourceEventToCoordinator(new BinlogSplitAssignedEvent());
             }
         }
         // notify split enumerator again about the finished unacked snapshot splits

--- a/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java
+++ b/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java
@@ -173,6 +173,20 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
     }
 
     @Test
+    public void testNewlyAddedTableForExistsPipelineTwiceWithAheadBinlogAndAutoCloseReader()
+            throws Exception {
+        testNewlyAddedTableOneByOne(
+                DEFAULT_PARALLELISM,
+                ", 'scan.incremental.close-idle-reader.enabled'='true'",
+                FailoverType.NONE,
+                FailoverPhase.NEVER,
+                true,
+                "address_hangzhou",
+                "address_beijing",
+                "address_shanghai");
+    }
+
+    @Test
     public void testNewlyAddedTableForExistsPipelineThrice() throws Exception {
         testNewlyAddedTableOneByOne(
                 DEFAULT_PARALLELISM,
@@ -588,7 +602,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                     getStreamExecutionEnvironment(finishedSavePointPath, parallelism);
             StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
 
-            String createTableStatement = getCreateTableStatement(captureAddressTables);
+            String createTableStatement = getCreateTableStatement("", captureAddressTables);
             tEnv.executeSql(createTableStatement);
             tEnv.executeSql(
                     "CREATE TABLE sink ("
@@ -630,7 +644,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                     getStreamExecutionEnvironment(finishedSavePointPath, parallelism);
             StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
 
-            String createTableStatement = getCreateTableStatement(captureTablesThisRound);
+            String createTableStatement = getCreateTableStatement("", captureTablesThisRound);
             tEnv.executeSql(createTableStatement);
             tEnv.executeSql(
                     "CREATE TABLE sink ("
@@ -703,6 +717,23 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
             boolean makeBinlogBeforeCapture,
             String... captureAddressTables)
             throws Exception {
+        testNewlyAddedTableOneByOne(
+                parallelism,
+                "",
+                failoverType,
+                failoverPhase,
+                makeBinlogBeforeCapture,
+                captureAddressTables);
+    }
+
+    private void testNewlyAddedTableOneByOne(
+            int parallelism,
+            String sourceOptions,
+            FailoverType failoverType,
+            FailoverPhase failoverPhase,
+            boolean makeBinlogBeforeCapture,
+            String... captureAddressTables)
+            throws Exception {
 
         // step 1: create mysql tables with initial data
         initialAddressTables(getConnection(), captureAddressTables);
@@ -727,7 +758,8 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                     getStreamExecutionEnvironment(finishedSavePointPath, parallelism);
             StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
 
-            String createTableStatement = getCreateTableStatement(captureTablesThisRound);
+            String createTableStatement =
+                    getCreateTableStatement(sourceOptions, captureTablesThisRound);
             tEnv.executeSql(createTableStatement);
             tEnv.executeSql(
                     "CREATE TABLE sink ("
@@ -836,7 +868,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
         }
     }
 
-    private String getCreateTableStatement(String... captureTableNames) {
+    private String getCreateTableStatement(String otherOptions, String... captureTableNames) {
         return format(
                 "CREATE TABLE address ("
                         + " table_name STRING METADATA VIRTUAL,"
@@ -858,6 +890,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                         + " 'server-time-zone' = 'UTC',"
                         + " 'server-id' = '%s',"
                         + " 'scan.newly-added-table.enabled' = 'true'"
+                        + " %s"
                         + ")",
                 MYSQL_CONTAINER.getHost(),
                 MYSQL_CONTAINER.getDatabasePort(),
@@ -865,7 +898,8 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                 customDatabase.getPassword(),
                 customDatabase.getDatabaseName(),
                 getTableNameRegex(captureTableNames),
-                getServerId());
+                getServerId(),
+                otherOptions);
     }
 
     private StreamExecutionEnvironment getStreamExecutionEnvironment(

--- a/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java
+++ b/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/NewlyAddedTableITCase.java
@@ -175,9 +175,11 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
     @Test
     public void testNewlyAddedTableForExistsPipelineTwiceWithAheadBinlogAndAutoCloseReader()
             throws Exception {
+        Map<String, String> otherOptions = new HashMap<>();
+        otherOptions.put("scan.incremental.close-idle-reader.enabled", "true");
         testNewlyAddedTableOneByOne(
                 DEFAULT_PARALLELISM,
-                ", 'scan.incremental.close-idle-reader.enabled'='true'",
+                otherOptions,
                 FailoverType.NONE,
                 FailoverPhase.NEVER,
                 true,
@@ -602,7 +604,8 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                     getStreamExecutionEnvironment(finishedSavePointPath, parallelism);
             StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
 
-            String createTableStatement = getCreateTableStatement("", captureAddressTables);
+            String createTableStatement =
+                    getCreateTableStatement(new HashMap<>(), captureAddressTables);
             tEnv.executeSql(createTableStatement);
             tEnv.executeSql(
                     "CREATE TABLE sink ("
@@ -644,7 +647,8 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                     getStreamExecutionEnvironment(finishedSavePointPath, parallelism);
             StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
 
-            String createTableStatement = getCreateTableStatement("", captureTablesThisRound);
+            String createTableStatement =
+                    getCreateTableStatement(new HashMap<>(), captureTablesThisRound);
             tEnv.executeSql(createTableStatement);
             tEnv.executeSql(
                     "CREATE TABLE sink ("
@@ -719,7 +723,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
             throws Exception {
         testNewlyAddedTableOneByOne(
                 parallelism,
-                "",
+                new HashMap<>(),
                 failoverType,
                 failoverPhase,
                 makeBinlogBeforeCapture,
@@ -728,7 +732,7 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
 
     private void testNewlyAddedTableOneByOne(
             int parallelism,
-            String sourceOptions,
+            Map<String, String> sourceOptions,
             FailoverType failoverType,
             FailoverPhase failoverPhase,
             boolean makeBinlogBeforeCapture,
@@ -868,7 +872,8 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
         }
     }
 
-    private String getCreateTableStatement(String otherOptions, String... captureTableNames) {
+    private String getCreateTableStatement(
+            Map<String, String> otherOptions, String... captureTableNames) {
         return format(
                 "CREATE TABLE address ("
                         + " table_name STRING METADATA VIRTUAL,"
@@ -899,7 +904,16 @@ public class NewlyAddedTableITCase extends MySqlSourceTestBase {
                 customDatabase.getDatabaseName(),
                 getTableNameRegex(captureTableNames),
                 getServerId(),
-                otherOptions);
+                otherOptions.isEmpty()
+                        ? ""
+                        : ","
+                                + otherOptions.entrySet().stream()
+                                        .map(
+                                                e ->
+                                                        String.format(
+                                                                "'%s'='%s'",
+                                                                e.getKey(), e.getValue()))
+                                        .collect(Collectors.joining(",")));
     }
 
     private StreamExecutionEnvironment getStreamExecutionEnvironment(

